protocol/0082-RFPR-on_chain_referral_program.md

```diff
@@ -0,0 +1,298 @@
+
+```pseudo
+team_epoch_volume = sum[min(party_epoch_volume, referralProgram.maxPartyVolumePerEpoch) for each party in team]
+```
+
+After the values are calculated, the `team_epoch_volume` is stored by the network.
+
+The network can then calculate the teams `team_running_volume` by summing a teams `team_epoch_volume` values over the last n epochs where n is the `window_length` set in the [governance proposal](#governance-proposals).
+
+## Benefit mechanics
+
+### Setting benefit factors
+
+Whilst a referral program is `STATUS_ACTIVE`, at the start of an epoch (after pending `JoinTeam` transactions have been processed) the network must set the `referral_reward_factor` and `referral_discount_factor` for each referee. This is done by identifying a referees team and identifying the teams current benefit tier. A teams benefit tier is defined as the highest tier for which their `team_running_volume` is greater or equal to the tiers `minimum_running_volume`. If a party does not qualify for any tier, both values are set to `0`.
+
+```pseudo
+Given:
+    benefit_tiers=[
+        {
+            "minimum_running_volume": 10000,
+            "referral_reward_factor": 0.001,
+            "referral_discount_factor": 0.001,
+        },
+        {
+            "minimum_running_volume": 20000,
+            "referral_reward_factor": 0.005,
+            "referral_discount_factor": 0.005,
+        },
+        {
+            "minimum_running_volume": 30000,
+            "referral_reward_factor": 0.010,
+            "referral_discount_factor": 0.010,
+        },
+    ]
+
+And:
+    team_running_volume=22353
+
+Then:
+    referral_reward_factor=0.005
+    referral_discount_factor=0.005
+```
+
+These benefit factors are then fixed for the duration of the next epoch.
+
+### Applying benefit factors
+
+Referral program benefit factors are applied by modifying [the fees](./0029-FEES-fees.md) paid by a party (either during continuous trading or on auction exit).
+
+
+## APIs
+
+The Teams API should expose the following information:
+
+- a list of all **teams** (by `id`), their founding **referrer**, and any current **referees**
+- a list of all **teams** (by `id`) and the following metrics:
+  - current `team_running_volume` (value at the start of the epoch)
+  - current `referral_reward_factor` applied to referee taker fees
+  - current `referral_discount_factor` applied to referee taker fees
+  - any data required for additional reward metrics
+
+The Trades API should now also expose the following additional information for every trade:
+
+- Referral program rewards
+  - `infrastructure_fee_reward`
+  - `liquidity_fee_reward`
+  - `maker_fee_reward`
+- Referral program discounts
+  - `infrastructure_fee_discount`
+  - `liquidity_fee_discount`
+  - `maker_fee_discount`
```

Comment:
Think this hopefully covers (1) from the [thread](https://vegaprotocol.slack.com/archives/CAHA5EX0F/p1690293264541239).